Refactor

aazuspan opened the pull request:
This is a large refactor of the sampling, training, and prediction workflows. The goal is to move most of the reusable code out of notebooks and into Python modules to clean things up. In the process, we should be able to add some additional flexibility to make it easier to predict different LiDAR attributes, train with different models and band combos, etc.

Very much in progress!

## TODO

- [x] Split out sampling into a Python module.
- [x] Refactor the sampling notebook.
- [x] Build data loading API. 
  - ~I was planning to use [tfio.IODataset.from_parquet](https://www.tensorflow.org/io/api_docs/python/tfio/experimental/IODataset#from_parquet) to load PNG byte strings and parse them back into images, but that is currently blocked by https://github.com/tensorflow/io/issues/1575. I may need to consider other options. Other possibilities are 1) eager loading Parquet with pandas (slow and won't scale well), 2) going back to TFRecords (complex), 3) using Dask dataframes (more complex), 4) trying another format with TFIO like Feather (worse compression), or 5) using image chips (potentially slow).~ Currently using HDF5 via `tfio.IODataset.from_hdf5`. This maintains the advantage of lazy evaluation, avoids the corrupted image issue with Parquet, and has the benefit of being able to store n-dimensional array features so that we don't need to encode our NAIP footprints into PNGs before storing them.
- [x] Refactor the modeling notebook for the new data API.
- [x] Refactor the inference workflow.

aazuspan commented:
@grovduck Just a heads up in case you wanted to follow along with progress, this is where I'll be focusing development for now.

I wrote up a quick [contributing guide](https://github.com/lemma-osu/naip-cnn/blob/main/CONTRIBUTING.md) (mostly so I don't forget how I got Docker working) that might be helpful if you feel like running things. The sampling module and notebook should be relatively stable now, but everything else is still deeply in progress!

Feel free to engage at whatever level of interest/amount of free time you have! I'm always open to feedback on design choices, but just knowing someone else might peruse the code should motivate me to write things a little more cleanly out of shame 😉

grovduck commented:
@aazuspan, definitely will want to follow along!  I'll try to get Docker set up on my machine and see what you've done here so far.  I'll try to provide some feedback here soon.  Thanks for including me!

aazuspan commented:
Awesome, let me know if you run into any issues with the setup. This is my first attempt at using Docker to manage a project, and I'm about 60% sure I'm doing it right...

grovduck commented:
Hey @aazuspan, stepping through the installation of Docker (all new to me) and there are a couple of places where I think some additional detail in `CONTRIBUTING.md` might help:

* On step 2 (installing Docker Desktop), I think it would be helpful for the user to know that Docker Desktop will go through an installation process (outside VSCode).  VSCode provides a message that this is going to happen, but it takes a while before the actual installation started for me.  Once installed, I had to log out of Windows and accept the subscription service agreement on restart.  Most people will likely figure this out, but being explicit might help.
* On step 3 (setting up WSL2 on Windows), it might be helpful to add a note about what the user should do if they are already using WSL, but Ubuntu is not their default distro.  This was my situation - I have a Debian distro (for the GNN download site) that was set as default and I switched it using `wsl --set-default Ubuntu`.  This is spelled out in the guide that you point to, so maybe not essential, but I had to think about this for a bit.
* On step 3 (and actually in the Docker Desktop instructions), it wasn't immediately clear (to Docker newbs like me) *where* you were supposed to run the `wsl` commands, through a normal Windows command shell or somehow through the Docker Desktop application.  Eventually, I figured out that it was the former and that the setup of WSL is outside of Docker Desktop.
* On step 5 (reopening the project in the container), it may be helpful for the user to know that the process can take a little while to complete.
* When checking the setup, it may freak users out to see that they are running as `root`.  I assume this is OK, but might be useful to let users know
* I'm pretty sure this last one is a corner case, but I actually have two GPUs on my machine, one of which is not compatible with CUDA, I believe.  When I run `assert tf.config.list_physical_devices('GPU')`, I'll get a warning that one of my GPUs is being ignored.  It looks like I can run `export CUDA_VISIBLE_DEVICES=<list of acceptable GPU device numbers>` at the linux prompt and it won't give me this warning
* Even when I do filter the GPUs, I still get warning messages that "Your kernel may have been built without NUMA support" on first run.  It looks like this may just be a warning that can be ignored, but I haven't dug too deep.

Please take all these suggestions with a grain of salt.  There probably won't be a lot of users trying to do this and it may not be worth the effort.  Most of these suggestions are extra "hand-holding" for the user.

grovduck reviewed (commented):
Hey @aazuspan, this is really great stuff.  Lots of new concepts and packages for me here, but you've written it clearly so far.  I bonked on the `02_training.ipynb` notebook when trying to read in the parquet data from the previous step, but it may just be a WIP at this point.  

Also, a general question.  You're using `hatch` here - I assume I should set that up within the container on Ubuntu?  Is that something that can be automated as part of the `Dockerfile`?  I've never worked with Docker files before, so newbie questions.

Let me know if/when you want me to continue reviewing.

aazuspan commented:
Thanks for the code review and the notes on the Docker setup! I'll start incorporating those. As you found out, things are pretty broken at the moment beyond the sampling stage, so I'll keep you posted as new pieces are ready for review.

aazuspan commented on notebooks/02_training.ipynb:
```diff
@@ -0,0 +1,202 @@
+    "NUM_EPOCHS = 20\n",
+    "BATCH_SIZE = 64\n",
+    "\n",
+    "full_dataset = load_parquet_dataset(\n",
+    "    \"../data/malheur_footprints_64x64.parquet\", \n",
+    "    target_names=(\"cancov\",), \n",
+    "    shape=(64, 64)\n",
+    ")\n",
```
Yes, that's exactly right. The `data` module was designed for my initial plan of storing columns for each band. I was updating it for the new encoded `image` column when I ran into that `tensorflow_io` bug, so that's all nonfunctional while I try and come up with a gameplan to move forward.

aazuspan commented on src/naip_cnn/data.py:
```diff
@@ -0,0 +1,36 @@
+            )
+        )
+    )
+    images = np.stack(images).astype(np.float32) / 255
+    labels = np.stack(labels).astype(np.float32) / 100
```
Yes. My current plan is to keep the package focused for this use case to avoid getting bogged down trying to maximize flexibility, so I'm thinking uint8 NAIP will probably be the only supported imagery. Also I'm not sure if this was clear, but I was planning to just keep the package as an internal tool to support the notebooks, rather than publishing it separately. What do you think - should we aim for something more flexible?

You're definitely right that the scaler for labels should be more general though, since we may be applying this to other metrics. Actually, I don't *think* the magnitude of the label affects the model (I just scaled it here because I was used to comparing MSEs at this magnitude), so we can probably ditch the scaling altogether.

grovduck replied:
All sounds good to me (private scope and not scaling the labels).

aazuspan commented on notebooks/01_sampling.ipynb:
```diff
@@ -0,0 +1,539 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We need a deterministic method for uniquely identifying our sample points so that we can potentially extract additional data in the future. We'll use [Plus Codes](https://maps.google.com/pluscodes/), which encode latitude and longitude coordinates into short strings."
```
Yeah, I've always thought they were cool but never had a use for them! I'm not sure how useful this *really* is since a sequential ID would have worked just as well so far, but it was a good excuse to bring them in.

aazuspan commented on src/naip_cnn/sampling.py:
```diff
@@ -0,0 +1,258 @@
+
+def extract_footprints_from_dataframe(
+    *,
+    df: pd.DataFrame,
```
Good catch, I think it does!

aazuspan commented on src/naip_cnn/sampling.py:
```diff
@@ -0,0 +1,258 @@
+    )
+
+
+def extract_footprints_from_dataframe(
```
Yeah, that's a good point. Were you thinking that should be specified in the function name or in the docstring?

aazuspan commented on src/naip_cnn/sampling.py:
```diff
@@ -0,0 +1,258 @@
+) -> ee.Geometry:
+    """Create a rectangular footprint centered on an xy coordinate."""
+    x, y = xy
+    minx = x - dims[0] // 2
```
Good call!

aazuspan commented on src/naip_cnn/sampling.py:
```diff
@@ -0,0 +1,258 @@
+        raise ValueError(msg)
+
+    data = raster.read(band)
+    band_name = raster.descriptions[band - 1]
```
No, I think you're right that that would break. I was vaguely thinking that `raster.descriptions` can't contain `None`, but that's probably not true if band names aren't set... Probably need to add a fallback here, maybe to `b1` or something similar.

aazuspan commented on src/naip_cnn/sampling.py:
```diff
@@ -0,0 +1,258 @@
+from PIL import Image
+
+
+def _create_point_grid(*, bounds, spacing: float, crs: str) -> gpd.GeoDataFrame:
```
Yeah, that was my thinking - do you think we might need to use it elsewhere in the project?

I'm a fan of the required kwargs too! I have a bad habit of calling my own functions with positional args, and got bit a few times when I changed the function signature without updating the calls.

grovduck replied:
> Yeah, that was my thinking - do you think we might need to use it elsewhere in the project?

Only in the sense that it's a useful function for other tasks (even outside this package).  I often struggle with where to keep these utility type functions that I keep writing over and over for different projects.  Not suggesting that you do this now, but do have thoughts on creating a separate "swiss army knife" package or what would be best practices on when you'd need this function again elsewhere?

aazuspan replied:
This is something I definitely run into as well, and so far I've always ended up re-implementing these utility functions again or copy-pasting between projects - definitely not the most efficient solution! I like the idea of a "swiss army knife" package, but I've never actually gone to the trouble of building one.

One of the reasons I've avoided it is that it would be difficult to integrate into public projects, e.g. we wouldn't want to include something like that it in sknnr because it would bloat the dependencies, and using it in published research would mean the package needs to be published as well. But there's no doubt it would be useful for internal projects, and if nothing else it would at least get those functions in one place if they need to be copied into other projects. I'm totally open to the idea of collaborating on a general purpose, geospatial data science package!

aazuspan commented on src/naip_cnn/sampling.py:
```diff
@@ -0,0 +1,258 @@
+        df.iloc[chunk_size * chunk : chunk_size * (chunk + 1)]
+        for chunk in range(n_chunks)
+    )
+    gdfs = Parallel(n_jobs=-1, prefer="threads")(
+        delayed(_extract_footprints_from_chunk)(chunk) for chunk in chunks
+    )
```
`joblib` is great! Definitely my goto for parallelizing.

aazuspan commented:
> Also, a general question. You're using hatch here - I assume I should set that up within the container on Ubuntu? Is that something that can be automated as part of the Dockerfile? I've never worked with Docker files before, so newbie questions.

At this point I don't think there's any need to use a Hatch environment, since we're already getting our isolated environment and dependencies from Docker. I used it because I figured we might want to add some Hatch scripts later, at which point we could add install it through the Dockerfile.

> On step 3 (setting up WSL2 on Windows), it might be helpful to add a note about what the user should do if they are already using WSL, but Ubuntu is not their default distro. This was my situation - I have a Debian distro (for the GNN download site) that was set as default and I switched it using wsl --set-default Ubuntu.

I actually only specified Ubuntu because that's what I used... You didn't happen to try Docker with your Debian install, did you? I imagine that would work, but not sure. In any case, explicitly recommending Ubuntu for consistency is probably a good idea.

> Even when I do filter the GPUs, I still get warning messages that "Your kernel may have been built without NUMA support" on first run. It looks like this may just be a warning that can be ignored, but I haven't dug too deep.

I get this too and came to the same conclusion, but also didn't dig too deep.

grovduck commented:
> Also, let me know if you have any thoughts/ideas on the topic of storing training data. I mostly moved away from `TFRecords` because I find them very confusing to work with and wanted to avoid the hassle of parsing data into and out of them, and my understanding is that their performance benefits are negligible at the scale we're likely to be working (or even negative as they have some overhead). I doubt the `tensorflow_io` issue is going to be resolved any time soon though, so I'm guessing we're going to need an alternative, whether that's another supported format, a different method for reading in Parquet, just storing chips, or going back to TFRecord.

You're way ahead of me on data formats - I barely know the `TFRecord` format and next to nothing about Parquet.  The only possible advantage to TFRecords that I can think of is the tight integration with GEE, but I think you're moving away from that possibly as well.  I thought the serialization of the arrays into PNG was pretty clever, so that would be great if that works, but I understand you may need to reevaluate based on the outstanding issue.

grovduck commented:
> You didn't happen to try Docker with your Debian install, did you?

I didn't try to set up with Debian, but I can do this if you'd like to test.  I think just using Ubuntu is fine though.

aazuspan commented:
@grovduck, I just pushed the updated data module and training notebook! I switched from Parquet to HDF5, which seems to be fully functional in `tfio` and has the added advantage of being able to store our NAIP arrays as-is without encoding, which saves a step when reading in datasets that should reduce training time a bit.

If you get a chance to run everything and train a model, I'm curious to hear if training time is much faster for you now. I'm still wrapping my head around Tensorflow datasets, but from the official examples I saw (e.g. [this one](https://www.tensorflow.org/datasets/keras_example)), I don't believe there's a reason to use `repeat(NUM_EPOCHS)` on the dataset as we were previously doing, so I removed that. That change caused each epoch to train for 375 steps (24,000 training samples // 64 samples per batch) instead of 1750 steps (375 batches * 20 epochs), which obviously runs much faster. I never ran training with the original TFRecord training data, so maybe that `repeat` was necessary with that solution.

Note that you'll need to re-run the sampling notebook to store the training data in the new format.

> The only possible advantage to TFRecords that I can think of is the tight integration with GEE

I didn't notice this until I started refactoring, but we weren't really taking advantage of the GEE support for TFRecords during sampling since we were just building the TFRecords client-side after grabbing the NAIP data from a collection of `ee.Dictionary` objects, so I don't think we lost any compatibility there.

We definitely may still want to use TFRecords for the mapping workflow, since GEE can chunk images into TFRecords for us, which might save some hassle over downloading GeoTIFFs and chunking ourselves (although I'm also curious to try [xbatcher](https://xbatcher.readthedocs.io/en/latest/)). If we ever wanted to operationalize the model and run inference in the cloud, I think you're right that there would be an advantage to using a TFRecord system for inference.

Also, just FYI, I experimented with the Planetary Computer idea that I floated and found it was MUCH slower than this GEE solution. Lesson learned!

> I think just using Ubuntu is fine though.

Agreed! I added a quick disclaimer to the contributing guide that other distros may work, but are untested.

grovduck commented:
> If you get a chance to run everything and train a model, I'm curious to hear if training time is much faster for you now

Yes, I did test both the sampling and training notebooks after you had made the HDF change and training was indeed much faster for me (the cell took 1m30s to evaluate - I think previously it was at least 15 minutes?).  

> I'm still wrapping my head around Tensorflow datasets, but from the official examples I saw (e.g. [this one](https://www.tensorflow.org/datasets/keras_example)), I don't believe there's a reason to use `repeat(NUM_EPOCHS)` on the dataset as we were previously doing, so I removed that.

This would definitely be a question for Adam.  Based on his original comment, he thought it was "mysterious" as well, but I think he was referencing some tutorials that he had found.  The only thing I've found is that `repeat` ensures that the model trains on the entire dataset for `NUM_EPOCHS`, but based on your explanation, it sounds like each epoch sees the full dataset with your change.

> We definitely may still want to use TFRecords for the mapping workflow, since GEE can chunk images into TFRecords for us

Yes, this is really what I meant, as I realized Adam didn't use GEE for the initial sampling scripts, but only during the inference step. `xbatcher` looks pretty interesting as well (I *really* need to understand `xarray` better!). 

> Also, just FYI, I experimented with the Planetary Computer idea that I floated and found it was MUCH slower than this GEE solution. Lesson learned!

Was this part of #1 or some other experimentation that you did?

aazuspan commented:
> I think he was referencing some tutorials that he had found. The only thing I've found is that repeat ensures that the model trains on the entire dataset for NUM_EPOCHS, but based on your explanation, it sounds like each epoch sees the full dataset with your change.

@sibley Any insights you can share on the use of `repeat` before training with the dataset?

> Was this part of https://github.com/lemma-osu/naip-cnn/issues/1 or some other experimentation that you did?

This was an earlier undocumented experiment. If I remember right, my Planetary Computer workflow took a few hours to pull the 5k chips that we can grab from EE in about a minute. There's probably a more efficient PC solution than what I came up with, but having to make a separate network request for each chip killed any speed advantage that came from being able to access the GeoTIFFs directly.

aazuspan commented:
I finally figured out the `repeat` question I was confused about, but thanks for the offer, Adam! I haven't dug back into the mapping code yet and I'm expecting to get confused by the TFRecord parsing, so I might reach out again for that AMA.

@grovduck if you're curious, it looks like `repeat` used in combination with `steps_per_epoch` used to be a necessary workaround due to an old Tensorflow bug. I found [this response](https://github.com/ageron/handson-ml2/issues/407#issuecomment-801777290) that sheds some light on the issue. It's talking about using an infinite `repeat`, but the same logic would apply to a finite `repeat`:

> ...using repeat(None) for train_set means that the training set will be an infinitely repeating dataset. Since we don't training to last forever, we must specify the number of steps_per_epoch and the number of epochs when calling the fit() method.

> A simpler way to do this nowadays is to avoid repeating the training set (don't call repeat() at all), then avoid using steps_per_epoch. Indeed Keras will train the model perfectly well: it will just go through the entire dataset at each epoch.
> The reason why I used the first approach in the book is that **there used to be a bug in Keras in TF 2.0 which made it impossible to use the simpler approach**: you just had to use the first approach.

We could keep the `repeat` and just specify `steps_per_epoch` when we `fit`, but it sounds like omitting both is now best practice.

grovduck commented on src/naip_cnn/data.py:
```diff
@@ -0,0 +1,160 @@
+        return ds.apply(tf.data.experimental.assert_cardinality(self.n))
+
+    def load_train(self, year: int = 2020, bands: tuple[str] = BANDS):
+        """Load a Tensorflow Dataset of training NAIP images from a Parquet file.
```
Change `Parquet` to  `HDF5`?  On `load_val` and `load_test` as well.

aazuspan replied:
Yes, good catch!

grovduck commented on src/naip_cnn/inference.py:
```diff
@@ -0,0 +1,42 @@
+
+def parse_imagery_tfrecord(
+    serialized_example: bytes,
+    bands: tuple[str] = ("R", "G", "B", "N"),
```
Does it make sense to import `BANDS` from `.config` and use here?

aazuspan replied:
Totally, I made the change!

grovduck reviewed (commented):
@aazuspan, this looks really great.  I ran the full series of notebooks and everything worked great.  Most of my comments are documentation related and most are just making it more explicit.  If you think the notebooks will be only used internally and by a crowd that can figure out the small details, feel free to ignore.

aazuspan commented on .dockerignore:
```diff
@@ -0,0 +1,3 @@
+data/**
+models/**
+.git/**
```
I'm actually not sure in practice... My understanding is that `data/**` explicitly ignores subdirectories, but it seems like just using `data` behaves the same way. It might just be an implicit vs. explicit difference.

aazuspan commented on .gitignore:
```diff
@@ -1,2 +1,4 @@
 *.pyc
-secrets/
\ No newline at end of file
+secrets/**
```
Currently I'm storing the GEE service account key and Google Cloud details in there, but they definitely don't need to be in the repo!

> Is there a way to store the key that GEE prompts you for in here?

Good question! The keys you enter in `ee.Authenticate` are very short-lived, but once you're authenticated it does store a longer-term key at `~/.config/earthengine`, which is where `ee.Initialize` looks. I experimented with copying that key into the Docker container, but found it was a little trickier than expected, especially if the key didn't already exist. I'm sure it could be done with a Bash script though. The other possible complication is that the container would always run with whatever EE account was authenticated locally, but I'm not sure that's really an issue. 

Having to authenticate every time is definitely a nuisance, so I think I'll open an issue to take another look at copying credentials into the container.

aazuspan commented on notebooks/04_inference.ipynb:
```diff
@@ -0,0 +1,239 @@
+    },
+    {
+     "data": {
+      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAjAAAAFcCAYAAADWL4TTAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjcuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8pXeV/AAAACXBIWXMAAA9hAAAPYQGoP6dpAAEAAElEQVR4nOz92Y9tW57Xi31GN5s1VxvNjtix23P2OXmyz6yGKorLBYxByJYsYck2j4j34qWeQLJASJZ45QH+CF78aq6ssq/hcrnUpQoqszLrZJ5+txE7utXOfozhh9+M2Fk09yY211VZjp90dHYXEWutOecYv/HtfirGGLmru7qru7qru7qru/oFKv3H/QLu6q7u6q7u6q7u6q7+c+u ...
```
Not quite sure I follow here. This can't happen if the user runs the notebook in order, can it?

grovduck replied:
Yeah, I'm not sure what I was thinking here.  Disregard this one!

aazuspan commented on notebooks/04_inference.ipynb:
```diff
@@ -0,0 +1,239 @@
+   "source": [
+    "### Load Metadata\n",
+    "\n",
+    "When Earth Engine images are exported as TFRecords, a sidecar JSON file is generated that includes the spatial reference metdata that we'll need in order to export predicted values back to a georeferenced raster image. We'll load and parse that first."
```
Agreed! I updated with some more specific instructions.

aazuspan commented on notebooks/03_export_naip.ipynb:
```diff
@@ -0,0 +1,132 @@
+   "source": [
+    "# 03: Export NAIP\n",
+    "\n",
+    "This notebook exports NAIP imagery from GEE for mapping."
```
Yes, much better!

I do suspect nobody else will ever read these notebooks, but it doesn't hurt to be explicit.

aazuspan commented on notebooks/02_training.ipynb:
Good comments! I tried to clarify the details you noted and removed the `history` for now.

grovduck commented:
LGTM!
